packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/subscribe.ts

```diff
@@ -0,0 +1,294 @@
+ */
+export interface ContainerMutatingProps {
+  /**
+   * The queue name and URLs to be added to the container environment.
+   */
+  readonly environment: { [key: string]: string };
+}
+
+/**
+ * This hook modifies the application container's environment to
+ * add the queue URLs for the subscription queues of the service.
+ */
+export class SubscribeMutatingHook extends ContainerMutatingHook {
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default SQS Queue, `eventsQueue`. It also supports creation of SQS Queue
+ * per SNS Topic and sets up the SNS Subscriptions accordingly.
+ *
+ * The created subscriptions can be accessed using `subscriptions` field and the
+ * default queue for this service can be accessed using the getter `<extension>.eventsQueue`.
+ */
+export class SubscribeExtension extends ServiceExtension {
+  /** The subscriptions created by the extension.
+   * @default undefined
+   */
+  public readonly subscriptions?: Subscription[];
+
+  private _eventsQueue!: sqs.IQueue;
+
+  private environment: { [key: string]: string } = {};
+
+  private deadLetterQueues: sqs.IQueue[] = [];
+
+  private props?: SubscribeProps;
+
+  constructor(props?: SubscribeProps) {
+    super('subscribe');
+
+    this.props = props;
+    if (props?.topicSubscriptions) {
+      this.subscriptions = [];
+    }
+
+    if (props?.eventsQueue && props?.eventsQueueProps) {
+      throw Error('You can only specify either a queue or queue props for creating the default events queue.');
+    }
+  }
+
+  /**
+   * This hook creates and sets the default queue `eventsQueue` and dead letter queue and other topic-specific
+   * according to the provided settings. It creates subscriptions for the topics and also adds the queue URLs to the
+   * environment.
+   * @param service The parent service which this extension has been added to
+   * @param scope The scope that this extension should create resources in
+   */
+  public prehook(service: Service, scope: Construct) {
+    this.parentService = service;
+    this.scope = scope;
+
+    if (this.props?.eventsQueue) {
+      this._eventsQueue = this.props.eventsQueue;
+    } else {
+      this._eventsQueue = this.createQueues('Events', this.props?.eventsQueueProps, this.props?.eventsDeadLetterQueueProps);
+    }
+    this.environment[`${this.parentService.id.toUpperCase()}_EVENTS_QUEUE_URL`] = this._eventsQueue.queueUrl;
+
+    if (this.props?.topicSubscriptions) {
+      const eventsQueueSubscriptions = [];
+      for (const topicSubscription of this.props.topicSubscriptions) {
```

Comment:
Optional: Given the amount of if/else logic in this for loop body you might want to consider moving that if/else tree out to its own private helper method inside the extension called something like _addTopicSubscription. It wouldn't be a public method for external customers to use, just a private method used internally to the extension, but it would help make this a bit easier to read by breaking down the big pieces

Reply:
Agree. Also since the logic is quite complicated here. Adding necessary comments might make it much easier for the other maintainers to read.